### PR TITLE
Add NFC permission support

### PR DESCRIPTION
**Pull request by Honry**
Make NDEFReader.scan() as a switch because obtaining
NFC permission should be triggered by user activation

### CODE REVIEW

**Comment by Honry** [comment]
@kenchris, @riju, PTAL, thanks! Preview: https://honry.github.io/webnfc-groceries/

**Comment by kenchris** [comment]
I would probably have added an NFC (the same one used by Chrome) icon instead that was grayed out when disabled.

Then also when there are no items yet then just add a background image (like when list is done) writing tap to start scanning

**Comment by kenchris** [comment]
A bit like:

image

**Comment by Honry** [comment]
@kenchris, that's pretty cool! Do we still need extra icon specific for enabling/disabling scan when there are already items?

**Comment by kenchris** [comment]
I don't think it harms

**Comment by Honry** [comment]
@kenchris, it's NFC icon for scanning now, PTAL again. Thanks! Preview: https://honry.github.io/webnfc-groceries/

**Comment by kenchris** [comment]
The icon button in the top bar should be an mwc-icon-button currently it looks way too big

**Comment by kenchris** [comment]
Also when I reload and it already has the permission, no reason to ask people to tap again

**Comment by Honry** [comment]
> Also when I reload and it already has the permission, no reason to ask people to tap again

It's odd, are you sure the permission prompts again? Tapping the icon is to enable the scan operation rather asking for permission.

**Comment by kenchris** [comment]
Not the permission prompt, but the background still asks me to touch to start scanning.

**Comment by kenchris** [comment]
I guess we are confusing with the wording :)

Maybe write "Tap here to start scanning your NFC tags" then later "Add item or touch an NFC tag"

**Comment by Honry** [comment]
Ah, I got your point, will fix it.

**Comment by Honry** [comment]
> Maybe write "Tap here to start scanning your NFC tags" then later "Add item or touch an NFC tag"

@kenchris, I add the later wording to the snack bar but still ask user to tap to scan when reload the page, as we treat scan operation as a switch via the nfc icon at the top bar. Does that make sense to you?

**Comment by kenchris** [comment]
I think that is the user already granted it once, we shouldn't ask again and just show the icon as white (turned on)

**Comment by kenchris** [comment]
Also disabled icon is gray, enabled one is white... tapping it should be able to turn off scanning if it is turned on

File: package.json [hunk 1]
```diff
@@ -31,7 +31,7 @@
     "@material/mwc-drawer": "^0.6.0",
     "@material/mwc-fab": "^0.6.0",
     "@material/mwc-formfield": "^0.6.0",
-    "@material/mwc-icon-button": "^0.6.0",
+    "@material/mwc-icon-button": "^0.11.1",
```
Comment (kenchris) [inline package.json:34]: btw for a future update it would be good to move to mwc-dialog etc instead of the community versions.

File: src/main-app.js [hunk 2]
```diff
@@ -74,6 +81,14 @@ export class AddDialog extends LitElement {
     } catch (err) {
       this._snackbar.close();
 
+      const permission = await navigator.permissions.query({ name: 'nfc' });
+      if (permission.state == 'denied' && err.name == "NotAllowedError") {
+        this._snackbar.labelText = "NFC permission is denied, please grant it in browser settings!";
+        this._actionBtn.textContent = "";
+        this._snackbar.open();
+        return;
```
Comment (kenchris) [inline src/main-app.js:90]: if we return here don't you miss adding some event listeners? I mean even if I grant in browser settings afterwards, will it work without reload?

File: src/main-app.js [hunk 3]
```diff
@@ -140,11 +159,47 @@ export class MainApplication extends LitElement {
       bottom: 16px;
       right: 16px;
     }
+
+    .btn-scan {
+      fill: #018786;
+    }
+
+    .disabled-scan {
```
Comment (kenchris) [inline src/main-app.js]: mwc-icon-button should already take care of these... or there is a toggle button.

https://mwc-demos.glitch.me/demos/icon-button-toggle.html

File: src/main-app.js [hunk 4]
```diff
@@ -140,11 +159,47 @@ export class MainApplication extends LitElement {
...
+      fill: lightgray;
+    }
+
+    .div-bg-scan {
```
Comment (kenchris) [inline src/main-app.js:166]: we also use similar CSS for centering the other background icon. Can't we use a class instead then?

File: src/main-app.js [hunk 5]
```diff
@@ -157,6 +212,36 @@ export class MainApplication extends LitElement {
       if (ev.detail.reason === "action") {
         window.location.reload();
       }
+      this._snackbar.labelText = "";
+      this._actionBtn.textContent = "";
+    });
+
+    this._store.addEventListener('change', ev => {
+      this._hideScanBg = true;
+      this.requestUpdate();
```
Comment (kenchris) [inline src/main-app.js]: Just make _hideScanBg a property?  `@property _showScanInfo`

File: src/main-app.js [hunk 6]
```diff
@@ -157,6 +212,36 @@ export class MainApplication extends LitElement {
...
+          if (data.product) {
+            this._store.set(data.product, data.description);
+          }
+        }
+      });
+    } catch(err) {
+      console.error("Reading NFC tags is not supported");
+    }
+
+    this._btnScan.addEventListener('click', async ev => {
```
Comment (kenchris) [inline src/main-app.js]: no need to add async as you don't use await inside.

Also you can probably easier add these in the html code `@click="${this._startScan}"

File: src/main-app.js [hunk 7]
```diff
@@ -174,6 +260,42 @@ export class MainApplication extends LitElement {
...
+    try {
+      if (this._btnScanSvg.classList.contains("disabled-scan")) {
+        await this._reader.scan(scanOption);
+        this._btnScanSvg.classList.remove("disabled-scan");
+        this._snackbar.labelText = "Add item or touch an NFC tag!";
+        this._actionBtn.textContent = "";
+        this._snackbar.open();
+        if (this._firstScan) {
+          this._hideScanBg = true;
+          this._firstScan = false;
```
Comment (kenchris) [inline src/main-app.js]: why two veriables, don't they mean the same?

File: src/main-app.js [hunk 8]
```diff
@@ -174,6 +260,42 @@ export class MainApplication extends LitElement {
...
+          this.requestUpdate();
+        }
+      } else {
+        const controller = new AbortController();
+        scanOption.signal = controller.signal;
+        // New invocation of scan() will cancel previous scan().
+        await this._reader.scan(scanOption);
+        controller.abort();
+        this._btnScanSvg.classList.add("disabled-scan");
+        this._snackbar.labelText = "NFC tags scan operation is disabled!";
```
Comment (kenchris) [inline src/main-app.js]: We should probably use . instead of ! in most places

File: src/main-app.js [hunk 9]
```diff
@@ -186,15 +308,29 @@ export class MainApplication extends LitElement {
           <mwc-top-app-bar>
             <mwc-icon-button slot="navigationIcon" icon="menu"></mwc-icon-button>
             <div slot="title">Groceries list</div>
+            <mwc-icon-button slot="actionItems" id="btnScan">
```
Comment (kenchris) [inline src/main-app.js]: we want the icon button toggle here :)

**Comment by Honry** [comment]
> Also disabled icon is gray, enabled one is white... tapping it should be able to turn off scanning if it is turned on

When we should hide the background icon? As long as there's already an item in the list?

**Comment by kenchris** [comment]
> When we should hide the background icon? As long as there's already an item in the list?

Sure like the other background icon. Above I was talking about the menu bar mwc-icon-button-toggle

File: src/main-app.js (hunk 2 above)
Comment (Honry) [reply src/main-app.js:90]: We can add an event listener to monitor permission change, but if the permission change is from push operation and user doesn't enable the scan operation before, should we enable the scan operation as well? i.e. we enable scan operation by default as long as nfc permission is granted.

Comment (kenchris) [reply src/main-app.js:90]: That sounds fine

File: src/main-app.js (hunk 4 above)
Comment (Honry) [reply src/main-app.js:166]: Yeah, they are similar, but where's the best place to export this class? Create a new js file?

File: src/main-app.js (hunk 2 above)
Comment (Honry) [reply src/main-app.js:90]: Thanks, done.

File: src/main-app.js (hunk 5 above)
Comment (Honry) [reply src/main-app.js]: Done.

File: src/main-app.js (hunk 7 above)
Comment (Honry) [reply src/main-app.js]: Now only one is sufficient. Done.

File: src/main-app.js (hunk 8 above)
Comment (Honry) [reply src/main-app.js]: Done.

File: src/main-app.js (hunk 9 above)
Comment (Honry) [reply src/main-app.js]: Done. :)

File: src/main-app.js (hunk 6 above)
Comment (Honry) [reply src/main-app.js]: Done.

File: package.json (hunk 1 above)
Comment (Honry) [reply package.json:34]: Sorry? I didn't get the point, we already use mwc-dialog here.

Comment (kenchris) [reply package.json:34]: Not from @material !

Comment (Honry) [reply package.json:34]: Do you mean use e.g. @authentic/mwc-icon-button instead in future?

Comment (kenchris) [reply package.json:34]: Move from authentic to material :) to the official ones

Comment (kenchris) [reply package.json:34]: so basically the other way around - but its a separate issue... just mentioned it as you were working on this.

**Comment by kenchris** [comment]
I see a bug when I open up the app, it shows the NFC background and the items !

**Comment by kenchris** [comment]
image

File: package.json (hunk 1 above)
Comment (Honry) [reply package.json:34]: Aha, got it now. :p Will update it in future.

**Comment by Honry** [comment]
Oops, I will figure it out.

**Comment by Honry** [comment]
@kenchris, bug fixed.

File: src/main-app.js [hunk 10]
```diff
@@ -211,9 +211,15 @@ export class MainApplication extends LitElement {
       this._actionBtn.textContent = "";
     });
 
-    this._store.addEventListener('change', ev => {
-      this._hideScanInfo = true;
-    }, { once: true });
+    const onchange = async () => {
+      for await (let entry of this._store.entries()) {
+        this._hideScanInfo = true;
+        this._store.removeEventListener('change', onchange);
```
Comment (kenchris) [inline src/main-app.js:217]: why is this removed multiple times?

Comment (kenchris) [reply src/main-app.js:217]: Why dont you not just modify this

image

just add the this._hideScanInfo = true as the first line of the for await loop

Comment (kenchris) [reply src/main-app.js:217]: like inside the if(first)

Comment (Honry) [reply src/main-app.js:217]: Looks like the scope of `_hideScanInfo` property is limited in main-app.js, that's why I have to add an `onchange` here.

Comment (Honry) [reply src/main-app.js:217]: > why is this removed multiple times?

Two scenarios here,
1. Reload page with existing cached store items, we invoke `onchange` directly to hide the scan info.
2. Load page with no cached store items, when new item is added, `change` event listener triggers to hide the scan info.

**Comment by kenchris** [comment]
I am thinking that a one time onboarding screen makes more sense now (in addition to the icon toggle always present)

**Comment by kenchris** [comment]
mio-design_assets_0B3wFuHgbfPkzY1BILVRnM01fdmc_communication-permissions-request1


Add groceries items by tapping on NFC stickers and the like.

[SKIP]  [>]

Then have next [>] open the permission prompt
